compiler/rustc_expand/src/mbe/macro_parser.rs

```diff
@@ -402,11 +419,21 @@ pub struct TtParser<'tt> {
 
     /// The set of mps that are waiting for the black-box parser.
     bb_mps: Vec<Box<MatcherPos<'tt>>>,
+
+    /// Pre-allocate an empty match array, so it can be cloned cheaply for macros with many rules
+    /// that have no metavars.
+    empty_matches: Lrc<NamedMatchVec>,
```

Comment:
Nit: perhaps it's better to move this optimization to a separate commit, the "Overhaul how matches are recorded" is already complex without it.

Reply:
Ok.